docs/adr/0001_cmsis_component_source_structure_recommendations.md

```diff
@@ -0,0 +1,119 @@
+# CMSIS component source structure recommendations
+Date: 2020-08-19
+
+## Status
+Proposed
+
+## Context
+
+[`ARM-software/CMSIS_5`](https://github.com/ARM-software/CMSIS_5) has been ported to Mbed OS. However, it is split across multiple directories and it is not immediately obvious that it comes from that directory. It also makes it difficult to create an independent CMake target can be built as a library that can be optionally be added by end users.
+
+The current proposal is to create a new directory `CMSIS_5/` at the root of the repository that will mirror closely the structure used by `ARM-software/CMSIS_5/` keeping only the directories and files needed by Mbed OS. It will also include Mbed OS specific files and rename some directories in order to work with Mbed OS build tools (see [Decision](#decision) section below). This will result in the removal of the `rtos/source/TARGET_CORTEX/` and `cmsis/` directories.
+
+Additionally it will provide the following advantages:
+* Better maintenance of component
+* Easy creation of an independent CMake target
+* Easy source navigation that mirrors closely the upstream repository
+
+
+## <a name="decision"></a>Decision
+The following directory restructure is recommended:
+
+```
++-- CMSIS_5/
+|   +-- README.md
+|   +-- docs/
+|   +-- CMSIS/
+|       +-- include/
+|          +-- cmsis/
+|               +-- mbed_cmsis_conf.h   # Currently in cmsis/
```

Comment:
This is a bit confusing as it's added by us. I know that we currently change the directory structure anyway, but in the future I would hope for 121 import. How do we make it obvious that's mbed specific/added config/headers?

Reply:
How about the following:
```
+-- CMSIS_5/
|   +-- README.md
|   +-- docs/
|   +-- CMSIS/
|       +-- device/
|           +-- include/
|               +-- mbed_cmsis_conf.h           # Currently in cmsis/
|           +-- RTE/
|               +-- include/
|                   +-- RTE_Components.h    # Currently in cmsis/
|       +-- RTOS2/
|           +-- Include/
|               +-- A                   # Content of rtos/source/TARGET_CORTEX/rtx5/Include/
...
```

Reply:
This is not sustainable when we will move to CMSIS-PAck integration as we are injecting the headers inside the external component. Can we do it in a way that in the future will allow us to just import the pack without adding anything on top? I know we still have patches and will need to sort this out as well, but would be good not to add more things that need to change. Maybe something like:
```
+-- mbed-os/
|   +-- cmsis/
|       +-- CMSIS_5/
|       +-- include/ or mbed/
|           
```

Reply:
Please review:
* https://github.com/ARMmbed/mbed-os/pull/13464/files#diff-30a452aba0af1f07e9628a3927396961
* https://github.com/ARMmbed/mbed-os/pull/13464/files#diff-f330f6f3e410acb388b9f474845d38b9